v1/event_format/json_formatter.cc

```diff
@@ -0,0 +1,159 @@
+        return absl::InvalidArgumentError("Required attribute in Cloud Event cannot be empty");
+    }
+
+    Json::Value root;
+    root[kCeIdKey.data()] = cloud_event.id();
+    root[kCeSourceKey.data()] = cloud_event.source();
+    root[kCeSpecKey.data()] = cloud_event.spec_version();
+    root[kCeTypeKey.data()] = cloud_event.type();
+
+    for (auto const& attr : cloud_event.attributes()) {
+        absl::StatusOr<Json::Value> json_printed = PrintToJson(attr.second);
+        if (!json_printed.ok()) {
+            return json_printed.status();
+        }
+        root[attr.first] = (*json_printed);
+    }
+
+    switch (cloud_event.data_oneof_case()) {
+        case CloudEvent::DataOneofCase::kBinaryData:
+            root[kBinaryDataKey.data()] = cloud_event.binary_data();
+            break;
+        case CloudEvent::DataOneofCase::kTextData:
+            root[kJsonDataKey.data()] = cloud_event.text_data();
+            break;
+        case CloudEvent::DataOneofCase::kProtoData:
+            // TODO (#17): Handle CloudEvent Any in JsonFormatter
+            return absl::UnimplementedError("protobuf::Any not supported yet.");
+        case CloudEvent::DATA_ONEOF_NOT_SET:
+            break;
+    }
+
+    // Write JSON serialization as std::string
+    Json::StreamWriterBuilder builder;
+    StructuredCloudEvent structured_ce;
+    structured_ce.format = Format::kJson;
+    structured_ce.serialization = Json::writeString(builder, root);
+    return structured_ce;
+}
+
+absl::StatusOr<CloudEvent> JsonFormatter::Deserialize(
+        const StructuredCloudEvent& structured_ce) {
+    // Validate that this is the right format to be handled by this object
+    if (structured_ce.format != Format::kJson) {
+        return absl::InternalError(
+            "This structured cloud event is not JSON formatted and should not be handled by a Json Formatter.");
+    }
+
+    // Create a Json::Value from the serialization string
+    std::string serialization = structured_ce.serialization;
+    Json::CharReaderBuilder builder;
+    Json::CharReader * reader = builder.newCharReader();
+    std::string errors;
+    Json::Value root;   
+
+    bool parsingSuccessful = reader->parse(
+        serialization.c_str(), 
+        serialization.c_str() + serialization.size(), 
+        &root, 
+        &errors);
+        
+    if (!parsingSuccessful) {
+        return absl::InvalidArgumentError(errors);
+    }
+
+
+    // check that serialization contains a valid CE
+    if (!root.isMember(kCeIdKey.data()) || 
+            !root.isMember(kCeSourceKey.data()) || 
+            !root.isMember(kCeSpecKey.data()) || 
+            !root.isMember(kCeTypeKey.data())) { // using isMember to avoid creating null member in JsonCpp
+        return absl::InvalidArgumentError(
+            "Provided input is missing required Cloud Event attributes.");
+    }
+
+    CloudEvent cloud_event;
+    
+    for (auto const& member : root.getMemberNames()) {
```

Comment:
Isn't 'time' of type Timestamp? I don't see it here.

Reply:
As I understand it, [json-binding spec](url) has URI and timestamp as string types.

So should I do a regex match on strings in json-serializations to infer and ce-types URI, URI-reference, and timestamp?